From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chenk008 <kongchen28@gmail.com>
Date: Thu, 18 Apr 2024 23:02:16 +0800
Subject: [PATCH] --EKS-PATCH-- prioritize user
 EtcdOptions.StorageConfig.StorageObjectCountTracker

Description:
* Fixes EtcdOptions.StorageObjectCountTracker is nil, APF estimator got ObjectCountNotFoundErr

Upstream PR, Issue, KEP, etc. links:
* This patch was initially based on cherry pick of Kubernetes commit a277aaf6cf93dd878c3e0e279fd77bd1c6b2613e
(https://github.com/kubernetes/kubernetes/pull/129469/commits/a277aaf6cf93dd878c3e0e279fd77bd1c6b2613e), which is part
of PR #124223 (https://github.com/kubernetes/kubernetes/pull/124223).

If this patch is based on an upstream commit, how (if at all) do this patch and the upstream source differ?
* N/A

If this patch's changes have not been added by upstream, why not?
* N/A

Other patches related to this patch:
* EKS-PATCH-Update_pkg/controlplane/apiserver/config_test.go.patch
* EKS-PATCH-Fix_StorageObjectCountTracker_is_nil,_apf_estimator_got_ObjectCountNotFoundErr.patch

Changes made to this patch after its initial creation and reasons for these changes:
* N/A

Kubernetes version this patch can be dropped:
* Likely 1.30.10 -- https://github.com/kubernetes/kubernetes/pull/129469
---
 staging/src/k8s.io/apiserver/pkg/server/options/etcd.go | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go b/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go
index a3516cf8805..10f9775efcc 100644
--- a/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go
@@ -396,7 +396,11 @@ func (f *StorageFactoryRestOptionsFactory) GetRESTOptions(resource schema.GroupR
 		EnableGarbageCollection:   f.Options.EnableGarbageCollection,
 		ResourcePrefix:            f.StorageFactory.ResourcePrefix(resource),
 		CountMetricPollPeriod:     f.Options.StorageConfig.CountMetricPollPeriod,
-		StorageObjectCountTracker: storageConfig.StorageObjectCountTracker,
+		StorageObjectCountTracker: f.Options.StorageConfig.StorageObjectCountTracker,
+	}
+
+	if ret.StorageObjectCountTracker == nil {
+		ret.StorageObjectCountTracker = storageConfig.StorageObjectCountTracker
 	}
 
 	if f.Options.EnableWatchCache {
